pmd-core/src/main/java/net/sourceforge/pmd/lang/antlr/AntlrRuleViolationFactory.java

```diff
@@ -0,0 +1,36 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.antlr;
+
+import net.sourceforge.pmd.Rule;
+import net.sourceforge.pmd.RuleContext;
+import net.sourceforge.pmd.RuleViolation;
+import net.sourceforge.pmd.lang.ast.AntlrBaseNode;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.rule.AbstractRuleViolationFactory;
+import net.sourceforge.pmd.lang.rule.ParametricRuleViolation;
+import net.sourceforge.pmd.lang.rule.RuleViolationFactory;
+
+public final class AntlrRuleViolationFactory extends AbstractRuleViolationFactory {
+    public static final RuleViolationFactory INSTANCE = new AntlrRuleViolationFactory();
+
+    private AntlrRuleViolationFactory() {
+    }
+
+    @Override
+    protected RuleViolation createRuleViolation(final Rule rule, final RuleContext ruleContext, final Node node,
+                                                final String message) {
+        return new ParametricRuleViolation<>(rule, ruleContext, (AntlrBaseNode) node, message);
+    }
+
+    @Override
+    protected RuleViolation createRuleViolation(final Rule rule, final RuleContext ruleContext, final Node node,
+                                                final String message, final int beginLine, final int endLine) {
+        final ParametricRuleViolation<AntlrBaseNode> violation = new ParametricRuleViolation<>(rule, ruleContext,
```

Comment:
I don't see the point of creating a `ParametricRuleViolation<AntlrBaseNode>` only to lose the preciser type by returning RuleViolation.

In any case I don't see why we need to extend the base class at all here. This is the 9th copy of basically the same implementation across the codebase... This is the fault of the absence of defaults I believe...

Ideally someone would fetch services only from a language instance, not by getting them statically. Eg `RuleViolationFactory rvf = LanguageRegistry.findLanguageByTerseName("swift").getRuleViolationFactory()`. Calling eg `new AntlrRuleViolationFactory()` or `AntlrRuleViolationFactory.getInstance()`couples the client code to this class which should basically be internal.

Reply:
I think I get what you are saying, but from what you are saying this abstraction is required all over PMD, we are not proud of making yet another duplication, but solving this problem would be a totally different task. 

With the base classes we are introducing here, at least we are guaranteeing that all future Antlr extensions won't be contributing to his issue and we will only need to refactor this class.